lib/src/shared/main/java/com/couchbase/lite/Blob.java

```diff
@@ -376,224 +411,169 @@ public String digest() {
+
+        C4BlobStore blobStore = null;
+        C4BlobKey key = null;
+        FLSliceResult res = null;
         try {
-            return new C4BlobKey(blobDigest);
+            blobStore = database.getBlobStore();
+
+            key = new C4BlobKey(blobDigest);
+
+            res = blobStore.getContents(key);
+
+            final byte[] newContent = res.getBuf();
+            if (newContent == null) { return null; }
+
+            cacheContent(newContent);
+
+            return newContent;
         }
         catch (LiteCoreException e) {
-            Log.w(DOMAIN, "Invalid digest: " + blobDigest, e);
-            return null;
+            final String msg = "Failed to obtain BlobStore content for digest: " + blobDigest;
+            Log.e(DOMAIN, msg, e);
+            throw new IllegalStateException(msg, e);
+        }
+        finally {
+            if (res != null) { res.free(); }
+            if (key != null) { key.free(); }
+            if (blobStore != null) { blobStore.free(); }
         }
     }
 
-    private byte[] getBytesFromInitialContentStream() {
-        // No recourse but to read the initial stream into memory:
-        final byte[] contentResult;
-        if (initialContentStream == null) {
-            throw new IllegalStateException("initialContentStream variable is null");
+    private C4BlobKey storeBytesInDatabase(C4BlobStore store) throws LiteCoreException, IOException {
+        final InputStream contentStream = getContentStream();
+        if (contentStream == null) {
+            throw new IllegalStateException(
+                "No data available to write for install."
+                    + "Please ensure that all blobs in the document have non-null content.");
         }
-        final ByteArrayOutputStream out = new ByteArrayOutputStream();
+
+        C4BlobWriteStream blobOut = null;
         try {
+            blobOut = store.openWriteStream();
+
             final byte[] buffer = new byte[MAX_CACHED_CONTENT_LENGTH];
-            try {
-                int bytesRead;
-                while ((bytesRead = initialContentStream.read(buffer)) >= 0) {
-                    out.write(buffer, 0, bytesRead);
-                }
-                contentResult = out.toByteArray();
-                blobLength = contentResult.length;
-                if (blobLength <= MAX_CACHED_CONTENT_LENGTH) {
-                    content = contentResult;  // cache for later re-use
-                }
-            }
-            catch (IOException e) {
-                Log.w(DOMAIN, "I/O Error with the given stream.", e);
-                throw new IllegalStateException(e);
-            }
-            finally {
-                try {
-                    initialContentStream.close();
-                    initialContentStream = null;
-                }
-                catch (IOException ignore) { }
+            this.blobLength = 0;
+
+            int n;
+            while ((n = contentStream.read(buffer)) > 0) {
```

Comment:
From Java doc:
```
the total number of bytes read into the buffer, or -1 if there is no more data because the end of the stream has been reached.
```

Is it safer to check for -1 or >= 0?

Reply:
I just corrected all of those comparisons to be the same.  The documentation says that you will either get an -1, or that at least one byte will be read.  Still, I think you are right, <= is better.